### PR TITLE
DEV2-3189 find symbols from user qurey in current file

### PR DESCRIPTION
**Pull request by yonip23**
1. if the user selected a code - use this selection.
2. else:
    1. divide the user query into words, and search each word as a symbol, using `vscode.executeWorkspaceSymbolProvider`.
    2. find the first symbol with the same URI as the current document, and use its code as the `selectedCode` field.
        this operation is cancelled if it takes more than 1s to complete, as we don't want to delay the response too much.

### CODE REVIEW

**Comment by yonip23** [comment]
TODO: bump chat version before merging

File: src/utils/execution.utils.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export async function executeTimeout<R>(
```
Comment (dimacodota) [inline src/utils/execution.utils.ts:1]: use rejectOnTimeout instead

File: src/tabnineChatWidget/handlers/getEditorContextHandler.ts [hunk 2]
```diff
@@ -58,6 +67,58 @@ export async function getEditorContext(): Promise<EditorContextResponse> {
   };
 }
 
+async function getSelectedCode(
+  editor: vscode.TextEditor,
+  request?: EditorContextRequest
+): Promise<string | undefined> {
+  const selectedCode = editor.document.getText(editor.selection);
+  if ((selectedCode && selectedCode !== "") || !request?.userQuery) {
```
Comment (dimacodota) [inline src/tabnineChatWidget/handlers/getEditorContextHandler.ts:75]: you can check on `editor.selection.isEmpty` instead

File: src/tabnineChatWidget/handlers/getEditorContextHandler.ts [hunk 3]
```diff
@@ -58,6 +67,58 @@ export async function getEditorContext(): Promise<EditorContextResponse> {
...
+    return selectedCode;
+  }
+
+  const wordsInQuery = request.userQuery.split(/\s+/gm).slice(0, 20);
```
Comment (dimacodota) [inline src/tabnineChatWidget/handlers/getEditorContextHandler.ts:79]: maybe something like 
``` 
const wordRegex = /\b\w+\b/g;
const matches = text.match(wordRegex);
```
would be more percise?

File: src/tabnineChatWidget/handlers/getEditorContextHandler.ts [hunk 4]
```diff
@@ -58,6 +67,58 @@ export async function getEditorContext(): Promise<EditorContextResponse> {
...
+        )
+        .catch((e) => {
+          Logger.debug(
+            `failed to fetch symbols for '${word}': ${(e as Error).message}`
+          );
+          return undefined;
+        })
+    )
+  );
+  const firstSymbol = allSymbols.find((s) => s?.length)?.find((s) => !!s);
```
Comment (dimacodota) [inline src/tabnineChatWidget/handlers/getEditorContextHandler.ts:115]: can you elaborate? why only the first symbol ?

File: src/tabnineChatWidget/handlers/getEditorContextHandler.ts (hunk 3 above)
Comment (yonip23) [reply src/tabnineChatWidget/handlers/getEditorContextHandler.ts:79]: which cases are you optimising with this regex? can you give an example of a prompt that requires this regex?

File: src/tabnineChatWidget/handlers/getEditorContextHandler.ts (hunk 4 above)
Comment (yonip23) [reply src/tabnineChatWidget/handlers/getEditorContextHandler.ts:115]: the selected code should be only one paragraph, so I figured lets go with the first occurrence of a symbol from the file.
What to you have in mind? to join all the occurrences?

Comment (dimacodota) [reply src/tabnineChatWidget/handlers/getEditorContextHandler.ts:116]: yep, tough to join them all... but we can start only with the first one

**Review by dimacodota — APPROVED** [review]
lgtm

**Pull request merged**